File: frontend/src/types/secret.ts
/**
 * Secret Types - TypeScript definitions for encrypted secrets system
 *
 * Mirrors backend Pydantic schemas for type safety
 */

/**
 * Secret scope - where it can be used
 */
export enum SecretScope {
  ORGANIZATION = 'organization', // Available across entire organization
  CHATBOT = 'chatbot',           // Only for specific chatbot
}

/**
 * Encryption provider used to encrypt the secret
 */
export enum EncryptionProvider {
  FERNET = 'fernet',   // Internal symmetric encryption (default)
  AWS_KMS = 'aws_kms', // AWS Key Management Service
  VAULT = 'vault',     // HashiCorp Vault
}

/**
 * Base secret fields (common to all schemas)
 */
export interface SecretBase {
  name: string;
  display_name: string;
  description?: string;
  scope: SecretScope;
  metadata?: Record<string, any>;
}

/**
 * Schema for creating a new secret
 */
export interface SecretCreate extends SecretBase {
  value: string; // Plaintext value to encrypt (never stored unencrypted)
  chatbot_id?: string; // Required if scope is CHATBOT
  encryption_provider?: EncryptionProvider;
  encryption_key_id?: string; // For AWS KMS ARN, Vault path, etc.
}

/**
 * Schema for updating a secret
 */
export interface SecretUpdate {
  display_name?: string;
  description?: string;
  value?: string; // New plaintext value (will be re-encrypted)
  is_active?: boolean;
  metadata?: Record<string, any>;
}

/**
 * Schema for rotating encryption key
 */
export interface SecretRotateKey {
  new_encryption_provider?: EncryptionProvider;
  new_key_id?: string;
}

/**
 * Secret in database (includes system fields, no decrypted value)
 * This is what the API returns in list/get endpoints
 */
export interface Secret extends SecretBase {
  id: string;
  organization_id: string;
  chatbot_id?: string;

  // Encryption configuration
  encryption_provider: EncryptionProvider;
  encryption_key_id?: string;
  encryption_metadata?: Record<string, any>;

  // Status
  is_active: boolean;

  // Audit
  last_used_at?: string; // ISO datetime
  usage_count: number;

  // Timestamps
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
}

/**
 * Secret validation response
 */
export interface SecretValidationResponse {
  is_valid: boolean;
  message?: string;
}

/**
 * Secret usage statistics
 */
export interface SecretUsageStats {
  secret_id: string;
  secret_name: string;
  display_name: string;
  usage_count: number;
  last_used_at?: string;
  created_at: string;
  is_active: boolean;
}

/**
 * List response with pagination info
 */
export interface SecretListResponse {
  items: Secret[];
  total: number;
}

/**
 * Filter options for listing secrets
 */
export interface SecretListFilters {
  chatbot_id?: string;
  scope?: SecretScope;
  is_active?: boolean;
}

/**
 * Props for SecretSelector component
 */
export interface SecretSelectorProps {
  value?: string; // Selected secret ID
  onChange: (secretId: string | undefined) => void;
  chatbot_id?: string; // Filter by chatbot
  scope?: SecretScope; // Filter by scope
  placeholder?: string;
  required?: boolean;
  disabled?: boolean;
  className?: string;
}

/**
 * Props for SecretModal component
 */
export interface SecretModalProps {
  isOpen: boolean;
  onClose: () => void;
  onSuccess: () => void;
  secret?: Secret; // If editing existing secret
  chatbot_id?: string; // Pre-fill chatbot_id for chatbot-scoped secrets
}

File: backend/app/graphql/types/secret.py
"""
Secret GraphQL Types
"""

from datetime import datetime
from typing import Optional

import strawberry


# ============================================
# SECRET TYPES
# ============================================

@strawberry.enum
class SecretScopeEnum:
    """Scope of the secret"""

    ORGANIZATION = "organization"
    CHATBOT = "chatbot"


@strawberry.enum
class EncryptionProviderEnum:
    """Encryption provider"""

    FERNET = "fernet"
    AWS_KMS = "aws_kms"
    VAULT = "vault"


@strawberry.type
class SecretType:
    """
    Secret type for GraphQL

    Note: NEVER includes the encrypted or decrypted value in list responses
    """

    id: strawberry.ID
    organization_id: strawberry.ID
    chatbot_id: Optional[strawberry.ID] = None
    name: str
    display_name: str
    description: Optional[str] = None
    scope: SecretScopeEnum
    encryption_provider: EncryptionProviderEnum
    encryption_key_id: Optional[str] = None
    encryption_metadata: Optional[strawberry.scalars.JSON] = None
    is_active: bool
    secret_metadata: Optional[strawberry.scalars.JSON] = None
    last_used_at: Optional[datetime] = None
    usage_count: int
    created_at: datetime
    updated_at: datetime


@strawberry.type
class SecretWithValueType:
    """
    Secret with decrypted value

    WARNING: Contains plaintext secret value
    Only used when explicitly retrieving a secret's value
    """

    id: strawberry.ID
    organization_id: strawberry.ID
    chatbot_id: Optional[strawberry.ID] = None
    name: str
    display_name: str
    description: Optional[str] = None
    scope: SecretScopeEnum
    value: str  # Decrypted plaintext value
    is_active: bool
    created_at: datetime


@strawberry.input
class SecretCreateInput:
    """Input for creating secret"""

    name: str
    display_name: str
    description: Optional[str] = None
    value: str  # Plaintext value to encrypt
    scope: SecretScopeEnum = SecretScopeEnum.CHATBOT
    chatbot_id: Optional[strawberry.ID] = None
    encryption_provider: EncryptionProviderEnum = EncryptionProviderEnum.FERNET
    encryption_key_id: Optional[str] = None
    secret_metadata: Optional[strawberry.scalars.JSON] = None


@strawberry.input
class SecretUpdateInput:
    """Input for updating secret"""

    display_name: Optional[str] = None
    description: Optional[str] = None
    value: Optional[str] = None  # New plaintext value (will be re-encrypted)
    is_active: Optional[bool] = None
    secret_metadata: Optional[strawberry.scalars.JSON] = None


# ============================================
# STATISTICS TYPES
# ============================================

@strawberry.type
class SecretUsageStatsType:
    """Statistics about secret usage"""

    secret_id: strawberry.ID
    secret_name: str
    display_name: str
    usage_count: int
    last_used_at: Optional[datetime] = None
    created_at: datetime
    is_active: bool
